### PR TITLE
feat(phase-4b): taxonomy doc + phase_4b_default schema + parser (#184, #185)

### LINKED ISSUE #184: [#158 sub-A] REVIEW_POLICY.md § Phase 4b Triggers — five-class taxonomy
Sub-issue of #158 (Phase 4b proactive triggers).

## Scope

Doc-only. Add a "Triggers" subsection to `REVIEW_POLICY.md` § Phase 4b that codifies the five-class taxonomy from #158's empirical observation. This is the foundational doc that the .yml schema field (sibling sub-issue B), the classifier (sibling D), and the CLAUDE.md update (sibling C) all reference.

## Implementation plan

1. Locate `REVIEW_POLICY.md § Phase 4b: Manual CLI Fallback (Human Handoff)` section. After the existing fallback-trigger description, insert a new `### Phase 4b Triggers` subsection.

2. The subsection enumerates **five trigger classes** with the detection heuristic for each. Lift the structure from #158:

   1. **State-machine changes** — PRs introducing/modifying discriminated-union or finite-state-machine types driving UI or service behavior. Detection: diff includes new/modified types matching `type \w+ = \| \{ kind: "..."` or similar tagged-union patterns; OR PR body explicitly says "state machine."
   2. **Concurrency / transactional code** — PRs touching `runTransaction`, `Promise.all`, optimistic updates, subscription handling, CAS patterns. Detection: diff includes `runTransaction` / `setSnapshot` / `applyOptimistic` / `compare-and-set` keywords; OR modifies `**/transactions/**`, `**/concurrency/**` paths.
   3. **Prompt design / LLM contract changes** — PRs adding/modifying few-shot examples, system prompts, structured-output schemas, grounding rules. Detection: paths matching `**/prompts/**`, `**/.v[0-9]+.md`, or LLM-tool-call schema files.
   4. **Cross-cutting refactors** — PRs that change a contract used by ≥3 callers OR touch both the type layer and service layer in the same diff.
   5. **Validation / invariant-enforcement code** — PRs changing anything pinning a product invariant. Detection: paths matching `**/validation/**`, `**/security/**`; OR PR body explicit "invariant" mention.

3. After the five-class list, add a one-paragraph "Cycle-time budget" note (per #158): proactive 4b adds 30 min to a few hours per PR; acceptable for high-risk changes, corrosive for trivial ones; the taxonomy keeps that latency targeted.

4. Cross-link from the new subsection to:
   - `.github/review-policy.yml` `phase_4b_default` field (sub-issue B's deliverable)
   - `scripts/phase-4b-classifier.sh` (sub-issue D's deliverable) — for repos that want automated classification
   - The matchline empirical evidence in #158's body (which PRs caught what bugs)

## Acceptance

- [ ] `REVIEW_POLICY.md` § Phase 4b has a new `### Phase 4b Triggers` subsection
- [ ] All five trigger classes listed with detection heuristics
- [ ] Cycle-time budget paragraph included
- [ ] Cross-links to sibling deliverables present
- [ ] No behavioral changes to existing Phase 4b fallback flow — this is opt-in additive policy

## Out of scope

- The .yml schema field (sub-issue B) — this PR only documents the taxonomy
- The classifier script (sub-issue D)
- The CLAUDE.md operating-instruction update (sub-issue C)
- Migration note for existing consumers (sub-issue E)

Authoring-Agent: claude

### PR DESCRIPTION
**Pull request by nathanjohnpayne**
Closes #184 and #185. Folds in #188 (migration paragraph) since it's a 1-paragraph addition to the same `REVIEW_POLICY.md` § Phase 4b Triggers section sub-issue A creates — saves a follow-up doc PR.

This is **PR 1 of 3** for parent #158:
- This PR: #184 + #185 (+ #188 fold-in) — taxonomy doc + schema + parser. **No agent-behavior change yet.**
- Next PR: #186 — `scripts/phase-4b-classifier.sh` with 5 trigger detectors + tests
- Final PR: #187 — CLAUDE.md / AGENTS.md operating-instruction update (depends on this PR + the classifier)

### #184 — REVIEW_POLICY.md § Phase 4b Triggers (taxonomy doc)

New subsection codifying the five-class taxonomy from #158's empirical observation. Each class has detection heuristic + why-it-matters explanation:

1. State-machine changes (discriminated unions, FSMs)
2. Concurrency / transactional code
3. Prompt design / LLM contract changes
4. Cross-cutting refactors
5. Validation / invariant-enforcement code

Plus cycle-time-budget framing and a "Migration for existing consumers" paragraph (the latter is sub-issue #188's deliverable, folded in here).

### #185 — `phase_4b_default` schema field + parser

`.github/review-policy.yml` gets a new top-level `phase_4b_default` field. Three options: `fallback-only` / `complex-changes` / `always`. Mergepath itself defaults to `complex-changes`; existing consumers without the field default to `fallback-only` (no behavior change without explicit opt-in).

`scripts/codex-review-check.sh` gains a new `policy_field` reader (top-level scalar, mirroring the existing in-block `codex_field`). Reads + validates + exports `PHASE_4B_DEFAULT`.

### Tests

`scripts/ci/check_workflow_parsers` extended with 6 new tests for the `policy_field` reader (bare value, trailing comment, missing field, mixed fields, quoted value, end-to-end against the real config). 13 → 19 tests, all PASS.

## Why no agent-behavior change yet

The parser exports the value but the classifier (sub-issue #186) and CLAUDE.md operating instructions (sub-issue #187) are what ACT on it. Intermediate state (field set but classifier not yet shipped) is intentional — it lets the schema land first and propagate to downstream consumers without forcing them to also pull in a not-yet-ready classifier.

## Verification

- `bash -n scripts/codex-review-check.sh` ✓
- `scripts/ci/check_workflow_parsers`: 19/19 PASS
- End-to-end run of `codex-review-check.sh` against an existing PR: parser + new export work; gate logic unchanged

## Authoring-Agent

Authoring-Agent: claude

## Self-Review

- **Correctness:** `policy_field` mirrors `codex_field`'s awk shape exactly, minus the in-block guard (top-level fields don't have one). Quote-stripping + comment-stripping behavior identical between the two readers. Validation rejects unknown values with a clear pointer at the doctrinal subsection.
- **Regression risk:** zero. New top-level field; existing parsers don't read it (so old codex-review-check.sh behavior is unchanged for the existing scalars). Existing-consumer migration explicit: missing field → `fallback-only` (the existing implicit behavior).
- **Style:** matches existing comment-block conventions in `.github/review-policy.yml` (header + commented option list + canonical reference). New REVIEW_POLICY.md subsection uses the same heading hierarchy and detection-heuristic format already established.
- **Test coverage:** 6 new fixtures covering happy path, edge cases (quotes, comments, missing field), and an end-to-end check against the real config so a schema-doc typo can't pass while breaking the live read.
- **Security/dependency hygiene:** no new dependencies, no new env vars beyond `PHASE_4B_DEFAULT` (export is intentional — sub-issue #186's classifier consumes it).

## Propagation

After merge: propagates to all 7 downstream consumers via the standard sync flow. Each consumer that doesn't add the field stays on `fallback-only` (no change). Consumers that opt in to `complex-changes` will see effects once #186 (classifier) and #187 (CLAUDE.md update) land.

### CODE REVIEW

**Comment by nathanpayne-claude** [comment]
@codex review

File: scripts/codex-review-check.sh [hunk 1]
```diff
@@ -124,6 +124,42 @@ codex_field() {
...
 
+# Read a top-level (block-less) scalar field. Same shape as codex_field
+# but without the in-block check — used for fields like `phase_4b_default`
+# (#185) that live at the document root rather than inside `codex:` /
+# `coderabbit:`. Outputs the value or empty on miss.
+policy_field() {
+  local field=$1
+  [ -f "$CONFIG" ] || return 0
+  awk -v field="$field" '
+    $1 == field":" {
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply scripts/codex-review-check.sh]: Fixed in 8402727. Anchored policy_field to top-level via `/^[^[:space:]]/ && $1 == field":"` — only matches lines that start at column 1 (no indent). Added 2 regression tests in check_workflow_parsers: (1) nested same-named key under `codex:` plus a top-level twin → reads only the top-level value; (2) nested key with no top-level twin → returns empty. 21 tests now PASS (was 19).

File: REVIEW_POLICY.md [hunk 2]
```diff
@@ -272,6 +272,54 @@ Phase 4b is invoked when Phase 4a escalates to disagreement or runaway, times ou
...
+
+The `phase_4b_default` field in `.github/review-policy.yml` controls when 4b fires proactively:
+
+| Value | Behavior |
+|-------|----------|
+| `fallback-only` | Current behavior. 4b only on 4a unavailability/escalation/timeout. |
+| `complex-changes` | Run `scripts/phase-4b-classifier.sh` AFTER 4a clears; if any trigger matches, post the 4b handoff before merging. |
+| `always` | Skip the classifier; post the 4b handoff for every external-review-threshold PR. |
+
+**Default for new repos:** `complex-changes` (the empirically validated middle ground). **Default for existing repos** that haven't added the field: `fallback-only` (no behavior change without explicit opt-in — see [Migration for existing consumers](#migration-for-existing-consumers) below).
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply REVIEW_POLICY.md:289]: Fixed in 8402727. Added an "Implementation status" paragraph immediately under the table that states this PR ships only the policy/parser piece — runtime branching that acts on the field lands in #187 (CLAUDE.md/AGENTS.md), classifier itself in #186. Setting `complex-changes` or `always` is a no-op behaviorally until both merge. The table heading also updated to "Behavior (intended; full wiring lands in #186 + #187)" so the doc/code drift you flagged is now explicit upstream of the table.

**Comment by nathanpayne-claude** [comment]
@codex review

**Comment by nathanpayne-claude** [comment]
## External Review Required

**PR:** #189 — feat(phase-4b): taxonomy doc + phase_4b_default schema + parser (#184, #185)
**Branch:** `feat/phase-4b-triggers-taxonomy-184-185`
**Author Agent:** claude

### Summary

PR 1 of 3 for parent #158 (Phase 4b proactive triggers). Doc + schema + parser only — no agent-behavior change yet. The runtime branching that ACTS on `phase_4b_default` lands in #187 (CLAUDE.md update), and the classifier itself is in #190 (sub-issue #186). Per the doctrine the table now explicitly says: setting `complex-changes` or `always` is a no-op behaviorally until both merge.

R1 review (Codex App) found a P2 + CR Major. Both addressed in [8402727](https://github.com/nathanjohnpayne/mergepath/commit/8402727):
1. `policy_field` anchored to top-level keys via `/^[^[:space:]]/ && $1 == field":"` — nested same-named keys no longer leak into the lookup. Plus 2 regression tests.
2. REVIEW_POLICY.md "Implementation status" paragraph added immediately under the schema table making it explicit that this PR ships only the parse-and-export piece, with pointers to #186 + #187 for the rest.

### Focus Areas

1. **`policy_field` top-level anchoring** ([codex-review-check.sh:131-145](https://github.com/nathanjohnpayne/mergepath/blob/feat/phase-4b-triggers-taxonomy-184-185/scripts/codex-review-check.sh)). The `/^[^[:space:]]/ && $1 == field":"` guard requires zero leading whitespace AND the first whitespace-separated token to be `field:`. Worth confirming this catches every nested-collision case — e.g., a future `coderabbit:` block adding a `phase_4b_default:` key would correctly NOT match.

2. **REVIEW_POLICY.md doc/code drift framing.** The "Implementation status" paragraph + the table heading change ("Behavior (intended; full wiring lands in #186 + #187)") both signal mid-migration state. Worth confirming this is the right doctrinal shape — alternative would be to NOT ship the doc until #187 lands, but that fragments the parent #158 across more PRs.

3. **`phase_4b_default: complex-changes` on mergepath itself.** This PR sets mergepath's own value to `complex-changes`. The parser exports `PHASE_4B_DEFAULT=complex-changes` but nothing acts on it, so behavior is unchanged. Worth confirming that's the right intermediate state vs. shipping mergepath at `fallback-only` until the classifier lands and only flipping to `complex-changes` after.

### Observations from Internal Review

- 21 tests pass in `check_workflow_parsers` (was 19 pre-PR; added 2 nested-key regression tests).
- The new sub-issues #184/#185/#188 are folded into one PR rather than three separate PRs because they all touch the same `REVIEW_POLICY.md` § Phase 4b Triggers section + sibling `.github/review-policy.yml` field — saves 2 follow-up PRs.

### Suggested External Reviewer

nathanpayne-codex

### Rationale for External Review

Codex App `@codex review` retrigger after my r1 fix push timed out at 599s without re-reviewing. Same App flakiness pattern observed on prior PRs in this session (#171, #173, #179, #180, #181). Phase 4b for an independent CLI verdict on the r1 hardening fixes.

— Authoring agent: claude. Human relays to codex CLI session for review.

**Review by nathanpayne-codex — APPROVED** [review]
Approved on current head `8402727`.

Verified the prior review findings are addressed: `policy_field` is anchored to document-root keys, regression coverage now exercises nested same-named keys, and the Phase 4b trigger documentation explicitly states that this PR ships only policy/parser support while runtime wiring lands separately. The current-head CodeRabbit finding on `REVIEW_POLICY.md` is resolved, and the remaining Codex P2 thread is outdated with the fix present in this head.

Verification run locally on `8402727`:
- `git diff --check 5f9ae6a...HEAD`
- `bash -n scripts/codex-review-check.sh scripts/ci/check_workflow_parsers`
- all executable `scripts/ci/check_*` scripts: PASS, with the repo's existing doc-location/top-level warning output
- focused parser probe: nested `codex.phase_4b_default` is ignored and top-level `phase_4b_default` is read.

**Comment by nathanpayne-claude** [comment]
@nathanjohnpayne — this PR is blocked only on the `needs-external-review` label.

Per [REVIEW_POLICY.md § Agent prohibitions](https://github.com/nathanjohnpayne/mergepath/blob/main/REVIEW_POLICY.md), agents do not remove this label. When you're ready, clear it from any device:

- GitHub UI: Labels sidebar → click `x` on `needs-external-review`
- CLI: `gh pr edit 189 --remove-label needs-external-review` --repo nathanjohnpayne/mergepath

Auto-merge will fire as soon as the label is gone.

**Context:** codex CLI Phase 4b APPROVED on 8402727 (review #4215678307). Branch protection requires Label Gate green; removing this label clears it.

— posted by `scripts/request-label-removal.sh`
